stl/inc/type_traits

```diff
@@ -1483,140 +1483,197 @@ class reference_wrapper;
-                                                                                                                       \
-    template <class _Callable, class... _Types>                                                                        \
-    CONSTEXPR auto _CONCAT(NAME_PREFIX, invoke)(_Callable && _Obj, _Types && ... _Args) noexcept(                      \
-        noexcept(_CONCAT(NAME_PREFIX, _Invoker) < _Callable,                                                           \
-            _Types... > ::_Call(_STD forward<_Callable>(_Obj), _STD forward<_Types>(_Args)...)))                       \
-        ->decltype(_CONCAT(NAME_PREFIX, _Invoker) < _Callable,                                                         \
-            _Types... > ::_Call(                                                                                       \
-                _STD forward<_Callable>(_Obj), _STD forward<_Types>(_Args)...)) { /* INVOKE a callable object */       \
-        return _CONCAT(NAME_PREFIX, _Invoker)<_Callable, _Types...>::_Call(                                            \
-            _STD forward<_Callable>(_Obj), _STD forward<_Types>(_Args)...);                                            \
+// FUNCTION TEMPLATE invoke
+#pragma warning(push) // TRANSITION, DevCom-936696
+#pragma warning(disable : 28278) // Function '%s' appears with no prototype in scope
+enum class _Invoker_strategy {
+    _Functor,
+    _Pmf_object,
+    _Pmf_refwrap,
+    _Pmf_pointer,
+    _Pmd_object,
+    _Pmd_refwrap,
+    _Pmd_pointer
+};
+
+struct _Invoker_functor {
+    static constexpr _Invoker_strategy _Strategy = _Invoker_strategy::_Functor;
+
+    template <class _Callable, class... _Types>
+    static constexpr auto _Call(_Callable&& _Obj, _Types&&... _Args) noexcept(
+        noexcept(static_cast<_Callable&&>(_Obj)(static_cast<_Types&&>(_Args)...)))
+        -> decltype(static_cast<_Callable&&>(_Obj)(static_cast<_Types&&>(_Args)...)) {
+        return static_cast<_Callable&&>(_Obj)(static_cast<_Types&&>(_Args)...);
     }
+};
+
+struct _Invoker_pmf_object {
+    static constexpr _Invoker_strategy _Strategy = _Invoker_strategy::_Pmf_object;
+
+    template <class _Decayed, class _Ty1, class... _Types2>
+    static constexpr auto _Call(_Decayed _Pmf, _Ty1&& _Arg1, _Types2&&... _Args2) noexcept(
+        noexcept((static_cast<_Ty1&&>(_Arg1).*_Pmf)(static_cast<_Types2&&>(_Args2)...)))
+        -> decltype((static_cast<_Ty1&&>(_Arg1).*_Pmf)(static_cast<_Types2&&>(_Args2)...)) {
+        return (static_cast<_Ty1&&>(_Arg1).*_Pmf)(static_cast<_Types2&&>(_Args2)...);
+    }
+};
+
+struct _Invoker_pmf_refwrap {
+    static constexpr _Invoker_strategy _Strategy = _Invoker_strategy::_Pmf_refwrap;
+
+    template <class _Decayed, class _Ty1, class... _Types2>
+    static constexpr auto _Call(_Decayed _Pmf, _Ty1&& _Arg1, _Types2&&... _Args2) noexcept(
+        noexcept((static_cast<_Ty1&&>(_Arg1).get().*_Pmf)(static_cast<_Types2&&>(_Args2)...)))

```

Comment:
Forwarding is extraneous for the `reference_wrapper` argument: `.get()` doesn't care if it's an lvalue or rvalue. These could equivalently be `_Arg1.get()`.

Reply:
I agree that forwarding can be removed. I'm not sure why I did that in my original implementation. I think it was because `reference_wrapper` support was added by LWG-2219 implemented in VS 2017 15.0, and I used either pmf_object or pmf_pointer as a starting point. Both of those must forward (for pmf_object, the PMF might be sensitive to the object's constness and value category; for pmf_pointer, a smart pointer might have an `operator*` that's sensitive to the smart pointer's constness and value category). I think I just didn't analyze whether forwarding mattered for `reference_wrapper` (because the entire philosophy of the `invoke` overhaul was that forwarding everything down was the simplest and best approach). Note that I did do such analysis for PMFs/PMDs themselves, which is why they're passed as `_Decayed` here.

`reference_wrapper` is trivially copyable, and it's the size of a pointer, so passing it by value (same "decayed" technique) might be reasonable, and might be more consistent (instead of passing `_Ty1&&` and just not forwarding it). So I think I would additionally recommend changing this to pass `_Refwrap _Rw`.

Summary: a novice copies code with no understanding of what is going on. An expert copies code with full understanding of what is going on, but then gets lazy and doesn't bother to simplify things that are no longer necessary.